### PR TITLE
Fixes microsoft#21219 + some improvements

### PR DESCRIPTION
**Pull request by vedantmgoyal2009**
- [ ] Have you signed the [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/winget-pkgs)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/microsoft/winget-pkgs/pulls) for the same manifest update/change?
- [ ] Have you validated your manifest locally with `winget validate --manifest <path>`? 
- [ ] Have you tested your manifest locally with `winget install --manifest <path>`?
- [ ] Does your manifest conform to the [1.0 schema](https://github.com/microsoft/winget-cli/blob/master/doc/ManifestSpecv1.0.md)?

Note: `<path>` is the name of the directory containing the manifest you're submitting.

-----

### CODE REVIEW

**Comment by OfficialEsco** [comment]
I'll start testing this now, my last 60+ PRs used version 1 of the SimpleUpdate

**Comment by vedantmgoyal2009** [comment]
@OfficialEsco I have fixed the issue.

@Trenly If you do any changes to the script, please commit directly to my branch. It is very difficult for me to update the branch and resolve conflicts.

**Comment by Trenly** [comment]
> @OfficialEsco I have fixed the issue.
> 
> @Trenly If you do any changes to the script, please commit directly to my branch. It is very difficult for me to update the branch and resolve conflicts.

Rebasing can be extremely difficult. It is a good skill to learn though, is how to read through and resolve merge conflicts during a rebase. I'll try to avoid any big changes while you have open PR's. I also don't want us to get into a situation where we have a branch of a branch of a branch of a branch for creating and merging in new features though, so I can't always guarantee there won't be conflicts

File: Tools/YamlCreate.ps1 [hunk 1]
```diff
@@ -391,13 +390,24 @@ Function Read-WinGet-InstallerValues {
             exit 1
         } finally {
             Write-Host "Time taken: $((Get-Date).Subtract($start_time).Seconds) second(s)" -ForegroundColor Green
+            
             $InstallerSha256 = (Get-FileHash -Path $script:dest -Algorithm SHA256).Hash
-            $FileInformation = Get-AppLockerFileInformation -Path $script:dest | Select-Object Publisher | Select-String -Pattern '{[A-Z0-9]{8}-([A-Z0-9]{4}-){3}[A-Z0-9]{12}}'
-            $MSIProductCode = $FileInformation.Matches
+            
+            if ($script:dest.EndsWith('msix','CurrentCultureIgnoreCase') -or $script:dest.EndsWith('msixbundle','CurrentCultureIgnoreCase')) { $InstallerType = 'msix' }
```
Comment (Trenly) [inline Tools/YamlCreate.ps1:396]:
```suggestion
            if ($script:dest -match "\.msix(bundle){0,1}$") { $InstallerType = 'msix' }
```

```raw
PS C:\WINDOWS\system32> $testString = "app.Msixbundle"
PS C:\WINDOWS\system32> $testString -match "msix(bundle){0,1}$"
True
PS C:\WINDOWS\system32> $testString = "app.Msixbundles"
PS C:\WINDOWS\system32> $testString -match "msix(bundle){0,1}$"
False
PS C:\WINDOWS\system32> $testString = "app.mSIx"
PS C:\WINDOWS\system32> $testString -match "msix(bundle){0,1}$"
True
PS C:\WINDOWS\system32> $testString = "app.msix-extractor.exe"
PS C:\WINDOWS\system32> $testString -match "msix(bundle){0,1}$"
False
```

File: Tools/YamlCreate.ps1 [hunk 2]
```diff
@@ -391,13 +390,24 @@ Function Read-WinGet-InstallerValues {
...
+            elseif ($script:dest.EndsWith('msi','CurrentCultureIgnoreCase')) { $InstallerType = 'msi' }
+            elseif ($script:dest.EndsWith('appx','CurrentCultureIgnoreCase') -or $script:dest.EndsWith('appxbundle','CurrentCultureIgnoreCase')) { $InstallerType = 'appx' }
```
Comment (Trenly) [inline Tools/YamlCreate.ps1:398]:
```suggestion
            elseif ($script:dest -match "\.appx(bundle){0,1}$") { $InstallerType = 'appx' }
```

File: Tools/YamlCreate.ps1 [hunk 3]
```diff
@@ -391,13 +390,24 @@ Function Read-WinGet-InstallerValues {
...
+            elseif ($script:dest.EndsWith('zip','CurrentCultureIgnoreCase')) { $InstallerType = 'zip' }
+
+            if ($InstallerUrl -match [regex]('\bx64\b') -or $InstallerUrl -match [regex]('\bwin64\b') -or $InstallerUrl -match [regex]('\b64\b')) {$architecture = 'x64'}
```
Comment (Trenly) [inline Tools/YamlCreate.ps1:401]:
```suggestion
            if ($InstallerUrl -match "\b(x|win){0,1}64\b") {$architecture = 'x64'}
```

File: Tools/YamlCreate.ps1 [hunk 4]
```diff
@@ -391,13 +390,24 @@ Function Read-WinGet-InstallerValues {
...
+            elseif ($InstallerUrl -match [regex]('\bx86\b') -or $InstallerUrl -match ('\bwin32\b') -or $InstallerUrl -match [regex]('\bia32\b') -or $InstallerUrl -match [regex]('\b86\b')) {$architecture = 'x86'}
```
Comment (Trenly) [inline Tools/YamlCreate.ps1:402]:
```suggestion
            elseif ($InstallerUrl -match "\b((win|ia)32)|(x{0,1}86)\b") {$architecture = 'x86'}
```

File: Tools/YamlCreate.ps1 [hunk 5]
```diff
@@ -391,13 +390,24 @@ Function Read-WinGet-InstallerValues {
...
+            elseif ($InstallerUrl -match [regex]('\barm64\b') -or $InstallerUrl -match [regex]('\baarch64\b')) { $architecture = 'arm64' }
```
Comment (Trenly) [inline Tools/YamlCreate.ps1:403]:
```suggestion
            elseif ($InstallerUrl -match "\b(arm|aarch)64\b") { $architecture = 'arm64' }
```

File: Tools/YamlCreate.ps1 [hunk 6]
```diff
@@ -391,13 +390,24 @@ Function Read-WinGet-InstallerValues {
...
+            elseif ($InstallerUrl -match [regex]('\barm\b')) { $architecture = 'arm' }
+
+            $MSIProductCode = $(Get-AppLockerFileInformation -Path $script:dest | Select-Object Publisher | Select-String -Pattern '{[A-Z0-9]{8}-([A-Z0-9]{4}-){3}[A-Z0-9]{12}}').Matches
+            
             if ($script:SaveOption -eq '1' -and -not($script:dest.EndsWith('appx', 'CurrentCultureIgnoreCase') -or $script:dest.EndsWith('msix', 'CurrentCultureIgnoreCase') -or $script:dest.EndsWith('appxbundle', 'CurrentCultureIgnoreCase') -or $script:dest.EndsWith('msixbundle', 'CurrentCultureIgnoreCase'))) { Remove-Item -Path $script:dest }
```
Comment (Trenly) [inline Tools/YamlCreate.ps1:408]:
```suggestion
            if ($script:SaveOption -eq '1' -and -not($script:dest -match "\.(msix|appx)(bundle){0,1}$") { Remove-Item -Path $script:dest }
```

File: Tools/YamlCreate.ps1 [hunk 7]
```diff
@@ -413,27 +423,31 @@ Function Read-WinGet-InstallerValues {
...
-        Write-Host -ForegroundColor 'Red' $script:_returnValue.ErrorString() 
-        Write-Host -ForegroundColor 'Green' -Object '[Required] Enter the architecture. Options:' , @($Patterns.ValidArchitectures -join ', ')
-        $architecture = Read-Host -Prompt 'Architecture' | TrimString
-        if ($architecture -Cin @($Patterns.ValidArchitectures)) {
-            $script:_returnValue = [ReturnValue]::Success()
-        } else {
-            $script:_returnValue = [ReturnValue]::new(400, 'Invalid Architecture', "Value must exist in the enum - $(@($Patterns.ValidArchitectures -join ', '))", 2)
-        }
-    } until ($script:_returnValue.StatusCode -eq [ReturnValue]::Success().StatusCode)
+    if (-not $architecture -Cin @($Patterns.ValidArchitectures)) {
```
Comment (Trenly) [inline Tools/YamlCreate.ps1:426]:
```suggestion
    if ($architecture -CNotIn @($Patterns.ValidArchitectures)) {
```

File: Tools/YamlCreate.ps1 [hunk 8]
```diff
@@ -413,27 +423,31 @@ Function Read-WinGet-InstallerValues {
...
-        Write-Host -ForegroundColor 'Red' $script:_returnValue.ErrorString() 
-        Write-Host -ForegroundColor 'Green' -Object '[Required] Enter the InstallerType. Options:' , @($Patterns.ValidInstallerTypes -join ', ' )
-        $InstallerType = Read-Host -Prompt 'InstallerType' | TrimString
-        if ($InstallerType -Cin @($Patterns.ValidInstallerTypes)) {
-            $script:_returnValue = [ReturnValue]::Success()
-        } else {
-            $script:_returnValue = [ReturnValue]::new(400, 'Invalid Installer Type', "Value must exist in the enum - $(@($Patterns.ValidInstallerTypes -join ', '))", 2)
-        }
-    } until ($script:_returnValue.StatusCode -eq [ReturnValue]::Success().StatusCode)
+    if (-not $InstallerType -Cin @($Patterns.ValidInstallerTypes)) {
```
Comment (Trenly) [inline Tools/YamlCreate.ps1:439]:
```suggestion
    if ($InstallerType -CNotIn @($Patterns.ValidInstallerTypes)) {
```

File: Tools/YamlCreate.ps1 [hunk 9]
```diff
@@ -1311,6 +1323,11 @@ Function Submit-Manifest {
     Write-Host
     if ($PromptSubmit -eq '0') {
         switch ($Option) {
+            'QuickUpdateVerison' {
+                if ($script:LastVersion -lt $script:PackageVersion ) { $CommitType = 'New version' }
+                elseif ($script:PackageVersion -in $script:ExistingVersions) { $CommitType = 'Update' }
+                elseif ($script:LastVersion -gt $script:PackageVersion ) { $CommitType = 'Add version' }
+             }
             'New' {
                 if ( $script:OldManifestType -eq 'None' ) { $CommitType = 'New package' }
```
Comment (Trenly) [inline Tools/YamlCreate.ps1:1332]:
```suggestion
        switch -regex ($Option) {
            'New|QuickUpdateVerison' {
                if ( $script:OldManifestType -eq 'None' ) { $CommitType = 'New package' }
```
Since the behavior is the same as `new` and we know that there will always be old manifests, we can use a regex switch here

File: Tools/YamlCreate.ps1 [hunk 10]
```diff
@@ -1580,6 +1597,7 @@ Function Read-PreviousWinGet-Manifest-Yaml {
     }
 
     if (-not (Test-Path -Path "$AppFolder\..")) {
+        if ($script:Option -eq 'QuickUpdateVerison') { Write-Host -ForegroundColor Red "This option requires manifest of previous version of the package. If you want to create a new package, please select Option 1."; exit }
```
Comment (Trenly) [inline Tools/YamlCreate.ps1:1600]: I don't think you need this because of line 1626; Since the option is `QuickUpdateVersion` and not `New`, line 1626 forces there to be manifests from the previous version. If they don't exist, it will exit the script. Please correct me if I'm mistaken though.

Additionally, this code (if placed here) would incorrectly not-trigger for packages which have a beta version but no production version.
For example; This would allow the script to continue for `Test.Application` if the folder for `Test.Application.Beta` existed but there were no manifests for `Test.Application`
```suggestion

```

**Review by Trenly — CHANGES_REQUESTED** [review]

File: Tools/YamlCreate.ps1 [hunk 11]
```diff
@@ -391,13 +390,24 @@ Function Read-WinGet-InstallerValues {
...
+            
             $InstallerSha256 = (Get-FileHash -Path $script:dest -Algorithm SHA256).Hash
-            $FileInformation = Get-AppLockerFileInformation -Path $script:dest | Select-Object Publisher | Select-String -Pattern '{[A-Z0-9]{8}-([A-Z0-9]{4}-){3}[A-Z0-9]{12}}'
-            $MSIProductCode = $FileInformation.Matches
-            if ($script:SaveOption -eq '1' -and -not($script:dest.EndsWith('appx', 'CurrentCultureIgnoreCase') -or $script:dest.EndsWith('msix', 'CurrentCultureIgnoreCase') -or $script:dest.EndsWith('appxbundle', 'CurrentCultureIgnoreCase') -or $script:dest.EndsWith('msixbundle', 'CurrentCultureIgnoreCase'))) { Remove-Item -Path $script:dest }
+            
+            if ($script:dest -match "\.msix(bundle){0,1}$") { $InstallerType = 'msix' }
+            elseif ($script:dest.EndsWith('msi','CurrentCultureIgnoreCase')) { $InstallerType = 'msi' }
+            elseif ($script:dest -match "\.appx(bundle){0,1}$") { $InstallerType = 'appx' }
+            elseif ($script:dest.EndsWith('zip','CurrentCultureIgnoreCase')) { $InstallerType = 'zip' }
```
Comment (Trenly) [inline Tools/YamlCreate.ps1:399]:
```suggestion
            elseif ($script:dest -match "\.msi$") { $InstallerType = 'msi' }
            elseif ($script:dest -match "\.appx(bundle){0,1}$") { $InstallerType = 'appx' }
            elseif ($script:dest -match "\.zip$") { $InstallerType = 'zip' }
```
Looks like I missed these; Forgot that this would trigger on things like `Package.fakemsi`

**Review by Trenly — COMMENTED** [review]
Sorry I missed these!

**Comment by vedantmgoyal2009** [comment]
@Trenly I committed all suggestions but then read your message.
image
So, if you comment the line, you will find that if I try to create a non-existing (new) package, it proceeds to enter InstallerUrl and the rest of the data. But, 
image
the script will throw error and exit.

The option is meant for only update the existing package version, so it is must that package should exist before selecting that option.

**Comment by vedantmgoyal2009** [comment]
https://github.com/Trenly/winget-pkgs/pull/23#issuecomment-912910428 @Trenly can only do. I have tried again but my brain got fully drained 🤯.

**Comment by Trenly** [comment]
> @Trenly I committed all suggestions but then read your message.
> image
> So, if you comment the line, you will find that if I try to create a non-existing (new) package, it proceeds to enter InstallerUrl and the rest of the data. But, if you uncomment that line, and the package does not exist before...
> image
> the script will throw error and exit.
> 
> The option is meant for only update the existing package version, so it is must that package should exist before selecting that option.
Ohhh, I see. So it needs to be in both places. I misunderstood

**Comment by Trenly** [comment]
> [#23 (comment)](https://github.com/Trenly/winget-pkgs/pull/23#issuecomment-912910428) @Trenly you can only do. I have tried again but my brain got fully drained 🤯.

What exactly isn't working about the multi-manifest Simple Update?

**Comment by vedantmgoyal2009** [comment]
Esco wants more fewer prompts. https://github.com/Trenly/winget-pkgs/pull/23#issuecomment-909674068

**Comment by vedantmgoyal2009** [comment]
Also, I have another feature that can be implemented pretty easily.

While creating manifests for packages like `Mirantis.Lens`, `Fndroid.ClashForWindows`, etc. we have to enter one URL two times to create different installers due to different scope/installer switches.

So, instead of downloading the same URL two times, it should check if that URL was entered before and if the condition is true, get the previous hash and continue to the next field.

**Comment by Trenly** [comment]
> @OfficialEsco @Trenly I have another feature that can be implemented pretty easily.
> 
> While creating manifests for packages like `Mirantis.Lens`, `Fndroid.ClashForWindows`, etc. we have to enter one URL two times to create different installers due to different scope/installer switches.
> 
> So, instead of downloading the same URL two times, it should check if that URL was entered before and if the condition is true, get the previous hash and continue to the next field.

Sounds like a good idea; Lets make a separate issue / pr for it when we do implement it

**Comment by Trenly** [comment]
> Esco wants more fewer prompts. [#23 (comment)](https://github.com/Trenly/winget-pkgs/pull/23#issuecomment-909674068)
> 
> Edit: Original Issue: [microsoft#21219](https://github.com/microsoft/winget-pkgs/issues/21219)

@OfficialEsco @vedantmgoyal2009  I've created a PR over at Vedant's repo to make the prompts reduced the way Esco had in the original post

https://github.com/vedantmgoyal2009/winget-pkgs/pull/4

**Comment by vedantmgoyal2009** [comment]
image
I think the warning should be displayed before entering PackageIdentifier because if a user sees a warning and then exits and starts over, he will have enter pkg id and version again. Also, is there any way to switch the mode without exiting the script?

**Comment by Trenly** [comment]
> image
> I think the warning should be displayed before entering PackageIdentifier because if a user sees a warning and then exits and starts over, he will have enter pkg id and version again. Also, is there any way to switch the mode without exiting the script?

We can move the warning, that shouldn’t be too difficult;

I can look into switching the option when I move the warning

**Comment by vedantmgoyal2009** [comment]
Oh, I have done it already. Pushing a new commit in a few minutes.

**Comment by Trenly** [comment]
> Oh, I have done it already. Pushing a new commit in a few minutes.

Still pushing that commit, or want me to run one through quick?

**Comment by vedantmgoyal2009** [comment]
I was correcting a spelling mistake. It took me an hour to find it.

**Comment by vedantmgoyal2009** [comment]
> > And, in "another feature" I think we have to submit two PR, first of deleting the old version and then adding a new version.
> 
> No it does not because the URL is static
> 
> ```
> Read Previous Manifest
> Create New Manifest
> Delete old Manifest if InstallerURL is the same
> Push Commit
> Create PR
> ```
> 
> Or
> 
> ```
> Read Previous Manifest
> Create New Manifest
> Push Commit
> Check if InstallerURL matches previous version
> If match Delete old Manifest
> Push Commit
> Create PR
> ```

We have to do this part also?

**Comment by vedantmgoyal2009** [comment]
Can we move code from functions like `Test-Manifest`, `Submit-Manifest`, `Read-WinGet-MandatoryInfo`, `Read-PreviousWinGet-Manifest-Yaml`, `Show-OptionMenu` at the last as you have done in commit 1f614b5

**Comment by Trenly** [comment]
> > > And, in "another feature" I think we have to submit two PR, first of deleting the old version and then adding a new version.
> > 
> > 
> > No it does not because the URL is static
> > ```
> > Read Previous Manifest
> > Create New Manifest
> > Delete old Manifest if InstallerURL is the same
> > Push Commit
> > Create PR
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Or
> > ```
> > Read Previous Manifest
> > Create New Manifest
> > Push Commit
> > Check if InstallerURL matches previous version
> > If match Delete old Manifest
> > Push Commit
> > Create PR
> > ```
> 
> We have to do this part also?
> 
> Edit: I just now realised that does @wingetbot do the same thing while it is creating "Automatic Update" PRs?

We will have to add in the functionality to automatically delete old versions, yes, but we don't want to stack too many features into a single PR.  The reason for this is that it makes any single change less likely to break the code. It's much easier to roll back a single feature to undo a mistake than it is to roll back all the features in a large PR and figure out which one caused the error. A rule of thumb I like to follow is to have one issue per PR unless they are extremely closely related. 

I did create #32 so we can track working on the deletion of previous manifests; I don't believe it is necessary to complete to merge the quick updates in though, since Esco

**Comment by Trenly** [comment]
> Can we move code from functions like `Test-Manifest`, `Submit-Manifest`, `Read-WinGet-MandatoryInfo`, `Read-PreviousWinGet-Manifest-Yaml`, `Show-OptionMenu` at the last as you have done in commit [1f614b5](https://github.com/Trenly/winget-pkgs/commit/1f614b5e463cf0986065aa6dd6ceedca28ec6786)

Yes, we can move code outside of functions if it makes sense to do so; However, if you do this you have to be very careful to respect program flow. More than once I've moved something outside of a function thinking it would be fine, only to find out later that the program flow wasn't quite the way I thought it was.   

There are a few reasons that we would also want to use functions; 1) Repeated code; The test and submit functions are common to all of the options, so it makes the most sense to have them as a function; 2) Readability; Even if you can do something without a function, sometimes having a named function makes it easier for others to understand the intent of the code

Tl;dr - Yes, in some cases, we just should be careful about it

**Comment by vedantmgoyal2009** [comment]
@Trenly Quick Update does not take Installer Switches.
image

**Comment by Trenly** [comment]
> @Trenly Quick Update does not take Installer Switches from previous manifest.
> Edit: It takes Installer Switches but why winget displays a warning 🤔?

It is because winget sees they are empty strings; Even if the values are defined, if the strings are empty then they are handled as if they are not specified

**Comment by Trenly** [comment]
@OfficialEsco Some pretty big restructures in this one; If you have any issues with it, please let us know. I'd like to merge this in before I start working on #30

**Review by OfficialEsco — APPROVED** [review]
It actually feels like this is ready to go tbh

**Comment by Trenly** [comment]
> It actually feels like this is ready to go tbh
Sweet! I'll merge it in soon then

**Review by Trenly — APPROVED** [review]
